kjdoc/tour.md

```diff
@@ -538,13 +538,441 @@ This section describes KJ APIs that control process execution and low-level inte
+
+Calling `.then()`, like `.wait()`, consumes the original promise, as if it were "moved away". Ownership of the original promise is transferred into the new, derived promise. If you want to register multiple continuations on the same promise, you must fork it first (see below).
+
+If the continuation itself returns another `Promise`, then the `Promise`s become chained. That is, the final type is reduced from `Promise<Promise<T>>` to just `Promise<T>`.
+
+```c++
+kj::Promise<kj::Own<kj::AsyncIoStream>> connectPromise =
+    network.connect(addr);
+kj::Promise<kj::String> textPromise = connectPromise
+    .then([](kj::Own<kj::AsyncIoStream> stream) {
+  return stream->readAllText().attach(kj::mv(stream));
+});
+```
+
+If a promise rejects (throws an exception), then the exception propagates through `.then()` to the new derived promise, without calling the continuation. If you'd like to actually handle the exception, you may pass a second lambda as the second argument to `.then()`.
+
+```c++
+kj::Promise<kj::String> promise = network.connect(addr)
+    .then([](kj::Own<kj::AsyncIoStream> stream) {
+  return stream->readAllText().attach(kj::mv(stream));
+}, [](kj::Exception&& exception) {
+  return kj::str("connection error: ", exception);
+});
+```
+
+You can also use `.catch_(errorHandler)`, which is a shortcut for `.then(identityFunction, errorHandler)`.
+
+### `kj::evalNow()`, `kj::evalLater()`, and `kj::evalLast()`
+
+These three functions take a lambda as the parameter, and return the result of evaluating the lambda. They differ in when, exactly, the execution happens.
+
+```c++
+kj::Promise<int> promise = kj::evalLater([]() {
+  int i = doSomething();
+  return i;
+});
+```
+
+As with `.then()` continuations, the lambda passed to these functions may itself return a `Promise`.
+
+`kj::evalNow()` executes the lambda immediately -- before `evalNow()` even returns. The purpose of `evalNow()` is to catch any exceptions thrown and turn them into a rejected promise. This is often a good idea when you don't want the caller to have to handle both synchronous and asynchronous exceptions -- wrapping your whole function in `kj::evalNow()` ensures that all exceptions are delivered asynchronously.
+
+`kj::evalLater()` executes the lambda on a future turn of the event loop. This is equivalent to `kj::Promise<void>().then()`.
+
+`kj::evalLast()` arranges for the lambda to be called only once all other work queued to the event loop has completed (but before querying the OS for new I/O events). This can often be useful, for example, for batching. For example, if a program tends to make many small write()s to a socket in rapid succession, you might want to add a layer that collects the writes into a batch, then sends the whole batch in a single write from an `evalLast()`. This way, none of the bytes are significantly delayed, but they can still be coalesced.
```

Comment:
Torn on this one. It does sound sort of weird, but the described example is just one specific example of "batching".

Reply:
Yeah, both instances of the phrase do have a purpose. Could change the first instance to `e.g.`, but maybe it's better to just leave as-is.

Reply:
```suggestion
`kj::evalLast()` arranges for the lambda to be called only once all other work queued to the event loop has completed (but before querying the OS for new I/O events). This can often be useful e.g. for batching. For example, if a program tends to make many small write()s to a socket in rapid succession, you might want to add a layer that collects the writes into a batch, then sends the whole batch in a single write from an `evalLast()`. This way, none of the bytes are significantly delayed, but they can still be coalesced.
```